File: Medium/73TaskAssignment.js
function taskAssignment(k, tasks) {
  const pairedTasks = [];
  const taskDurationsToIndices = getTaskDurationToIndices(tasks);

  const sortedTasks = [...tasks].sort((a, b) => a - b);
  for (let idx = 0; idx < k; idx++) {
    const taskDuration = sortedTasks[idx];
    const indicesWithTaskDuration = taskDurationsToIndices[taskDuration];
    const task1Index = indicesWithTaskDuration.pop();

    const task2SortedIndex = tasks.length - 1 - idx;
    const task2Duration = sortedTasks[task2SortedIndex];
    const indicesWithTaskDuration2 = taskDurationsToIndices[task2Duration];
    const task2Index = indicesWithTaskDuration2.pop();

    pairedTasks.push([task1Index, task2Index]);
  }

  return pairedTasks;
}

function getTaskDurationToIndices(tasks) {
  const taskDurationsToIndices = {};

  for (let idx = 0; idx < tasks.length; idx++) {
    const taskDuration = tasks[idx];
    if (taskDuration in taskDurationsToIndices) {
      taskDurationsToIndices[taskDuration].push(idx);
    } else {
      taskDurationsToIndices[taskDuration] = [idx];
    }
  }

  return taskDurationsToIndices;
}

File: Medium/73TaskAssignment.py
def taskAssignment(k, tasks):
    pairedTasks = []
    taskDurationsToIndices = getTaskDurationToIndices(tasks)

    sortedTasks = sorted(tasks)
    for idx in range(k):
        taskDuration = sortedTasks[idx]
        indicesWithTaskDuration = taskDurationsToIndices[taskDuration]
        task1Index = indicesWithTaskDuration.pop()

        task2SortedIndex = len(tasks) - 1 - idx
        task2Duration = sortedTasks[task2SortedIndex]
        indicesWithTaskDuration = taskDurationsToIndices[task2Duration]
        task2Index = indicesWithTaskDuration.pop()

        pairedTasks.append([task1Index, task2Index])

    return pairedTasks


def getTaskDurationToIndices(tasks):
    taskDurationsToIndices = {}

    for idx, taskDuration in enumerate(tasks):
        if taskDuration in taskDurationsToIndices:
            taskDurationsToIndices[taskDuration].append(idx)
        else:
            taskDurationsToIndices[taskDuration] = [idx]

    return taskDurationsToIndices
